source/iopipe/stream.d

```diff
@@ -77,3 +79,114 @@ struct ZeroDev
 
 /// Common instance of ZeroDev to use anywhere needed.
 immutable ZeroDev zeroDev;
+
+// helper for copying data from a src range into a buffer
+private size_t copy(Src, Buf)(ref Src src, Buf buf) if(is(immutable(ElementEncodingType!Buf) == immutable(ElementEncodingType!Src)))
+{
+    // element compatible ranges. Use standard copying
+    static if(is(Buf : T[], T) && is(Src : U[], U) && is(immutable(T) == immutable(U)))
+    {
+        // both dynamic arrays, use slice assign.
+        auto n = src.length < buf.length ? src.length : buf.length;
+        buf[0 .. n] = src[0 .. n];
+        src = src[n .. $];
+        return n;
+    }
+    else
+    {
+        size_t n = 0;
+        while(n < buf.length && !src.empty)
+        {
+            buf[n++] = src.front;
+            src.popFront;
+        }
+        return n;
+    }
+}
+
+struct RangeInput(R) {
+    private import std.traits: isNarrowString;
+    private {
+        R src;
+        static if(is(ElementType!R == T[], T))
+        {
+            enum isRangeOfSlices = true;
+            T[] data;
+        }
+        else
+            enum isRangeOfSlices = false;
+    }
+
+    this(R src)
+    {
+        this.src = src;
+        static if(isRangeOfSlices)
+            if(!this.src.empty)
+                data = this.src.front;
+    }
+
+    size_t read(Buf)(Buf buf) if (isNarrowString!Buf || isRandomAccessRange!Buf)
+    {
+        static if(is(typeof(copy(src, buf))))
```

Comment:
Am I correct to think that this is mostly equivalent to `static if(__traits(compiles, ...))`?

Reply:
It is mostly equivalent. I probably should switch to what you wrote.